src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts

```diff
@@ -177,4 +179,15 @@ export class PoolListComponent implements OnInit {
 
     return strings.join(', ');
   }
+
+  getSelectionTiers() {
+    const cacheTierIds = this.selection.hasSingleSelection ? this.selection.first()['tiers'] : [];
+    this.selectionCacheTiers.splice(0, cacheTierIds.length);
```

Comment:
You should better reassign the array to another or clear the array completely, as this will cause a bug if you have one pool with multiple tiers and another with less tiers. If you first select the one with more tiers the other will have the same amount of tiers even if it in reality hasn't.

I have prepared some tests to illustrate that:
```diff
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.spec.ts b/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.spec.ts
index cae0475132..3dd5a6acf7 100644
--- a/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.spec.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.spec.ts
@@ -23,6 +23,7 @@ describe('PoolListComponent', () => {
   let component: PoolListComponent;
   let fixture: ComponentFixture<PoolListComponent>;
   let poolService: PoolService;
+  let pools: Pool[];
 
   configureTestBed({
     declarations: [PoolListComponent, PoolDetailsComponent],
@@ -36,11 +37,28 @@ describe('PoolListComponent', () => {
     providers: [i18nProviders, PgCategoryService]
   });
 
+  const addPool = (name, id) => {
+    const pool = new Pool(name);
+    pool.pool = id;
+    pool.pg_num = 256;
+    pools.push(pool);
+  };
+
+  const setUpPools = () => {
+    pools = [];
+    addPool('a', 0);
+    addPool('b', 1);
+    addPool('c', 2);
+    component.pools = pools;
+  };
+
   beforeEach(() => {
     fixture = TestBed.createComponent(PoolListComponent);
     component = fixture.componentInstance;
     component.permissions.pool.read = true;
     poolService = TestBed.get(PoolService);
+    setUpPools();
+    spyOn(poolService, 'getList').and.callFake(() => of(pools));
     fixture.detectChanges();
   });
 
@@ -95,15 +113,8 @@ describe('PoolListComponent', () => {
   });
 
   describe('handling of executing tasks', () => {
-    let pools: Pool[];
     let summaryService: SummaryService;
 
-    const addPool = (name) => {
-      const pool = new Pool(name);
-      pool.pg_num = 256;
-      pools.push(pool);
-    };
-
     const addTask = (name: string, pool: string) => {
       const task = new ExecutingTask();
       task.name = name;
@@ -114,13 +125,7 @@ describe('PoolListComponent', () => {
     beforeEach(() => {
       summaryService = TestBed.get(SummaryService);
       summaryService['summaryDataSource'].next({ executing_tasks: [], finished_tasks: [] });
-      pools = [];
-      addPool('a');
-      addPool('b');
-      addPool('c');
-      component.pools = pools;
-      spyOn(poolService, 'getList').and.callFake(() => of(pools));
-      fixture.detectChanges();
+      setUpPools();
     });
 
     it('gets all pools without executing pools', () => {
@@ -241,4 +246,46 @@ describe('PoolListComponent', () => {
       expect(component.transformPgStatus(pgStatus)).toEqual(expected);
     });
   });
+
+  describe('getSelectionTiers', () => {
+    const setSelectionTiers = (tiers: number[]) => {
+      component.selection.selected = [
+        {
+          tiers
+        }
+      ];
+      component.selection.update();
+      component.getSelectionTiers();
+    };
+
+    it('should select multiple existing cache tiers', () => {
+      setSelectionTiers([0, 1, 2]);
+      expect(component.selectionCacheTiers).toEqual(pools);
+    });
+
+    it('should select cache existing tier', () => {
+      setSelectionTiers([0]);
+      expect(component.selectionCacheTiers).toEqual([{ pg_num: 256, pool: 0, pool_name: 'a' }]);
+    });
+
+    it('should not select cache tier if id is invalid', () => {
+      setSelectionTiers([-1]);
+      expect(component.selectionCacheTiers).toEqual([]);
+    });
+
+    it('should not select cache tier if empty', () => {
+      setSelectionTiers([]);
+      expect(component.selectionCacheTiers).toEqual([]);
+    });
+
+    // This test will fail with the current solution
+    it('should be able to selected one pool with multiple tiers, than with a single tier, than with no tiers', () => {
+      setSelectionTiers([0, 1, 2]);
+      expect(component.selectionCacheTiers).toEqual(pools);
+      setSelectionTiers([0]);
+      expect(component.selectionCacheTiers).toEqual([{ pg_num: 256, pool: 0, pool_name: 'a' }]);
+      setSelectionTiers([]);
+      expect(component.selectionCacheTiers).toEqual([]);
+    });
+  });
 });
```

My solution to fix that would be the following:
```diff
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts b/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
index d27d198aec..af55f3a12a 100644
--- a/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
@@ -182,12 +182,6 @@ export class PoolListComponent implements OnInit {
 
   getSelectionTiers() {
     const cacheTierIds = this.selection.hasSingleSelection ? this.selection.first()['tiers'] : [];
-    this.selectionCacheTiers.splice(0, cacheTierIds.length);
-
-    this.pools.forEach((pool) => {
-      if (cacheTierIds.indexOf(pool.pool) > -1) {
-        this.selectionCacheTiers.push(pool);
-      }
-    });
+    this.selectionCacheTiers = this.pools.filter((pool) => cacheTierIds.includes(pool.pool))
   }
 }
```

Reply:
> You should better reassign the array to another or clear the array completely, as this will cause a bug if you have one pool with multiple tiers and another with less tiers. If you first select the one with more tiers the other will have the same amount of tiers even if it in reality hasn't.
> 
> I have prepared some tests to illustrate that:
> 
> ```diff
> diff --git a/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.spec.ts b/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.spec.ts
> index cae0475132..3dd5a6acf7 100644
> --- a/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.spec.ts
> +++ b/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.spec.ts
> @@ -23,6 +23,7 @@ describe('PoolListComponent', () => {
>    let component: PoolListComponent;
>    let fixture: ComponentFixture<PoolListComponent>;
>    let poolService: PoolService;
> +  let pools: Pool[];
>  
>    configureTestBed({
>      declarations: [PoolListComponent, PoolDetailsComponent],
> @@ -36,11 +37,28 @@ describe('PoolListComponent', () => {
>      providers: [i18nProviders, PgCategoryService]
>    });
>  
> +  const addPool = (name, id) => {
> +    const pool = new Pool(name);
> +    pool.pool = id;
> +    pool.pg_num = 256;
> +    pools.push(pool);
> +  };
> +
> +  const setUpPools = () => {
> +    pools = [];
> +    addPool('a', 0);
> +    addPool('b', 1);
> +    addPool('c', 2);
> +    component.pools = pools;
> +  };
> +
>    beforeEach(() => {
>      fixture = TestBed.createComponent(PoolListComponent);
>      component = fixture.componentInstance;
>      component.permissions.pool.read = true;
>      poolService = TestBed.get(PoolService);
> +    setUpPools();
> +    spyOn(poolService, 'getList').and.callFake(() => of(pools));
>      fixture.detectChanges();
>    });
>  
> @@ -95,15 +113,8 @@ describe('PoolListComponent', () => {
>    });
>  
>    describe('handling of executing tasks', () => {
> -    let pools: Pool[];
>      let summaryService: SummaryService;
>  
> -    const addPool = (name) => {
> -      const pool = new Pool(name);
> -      pool.pg_num = 256;
> -      pools.push(pool);
> -    };
> -
>      const addTask = (name: string, pool: string) => {
>        const task = new ExecutingTask();
>        task.name = name;
> @@ -114,13 +125,7 @@ describe('PoolListComponent', () => {
>      beforeEach(() => {
>        summaryService = TestBed.get(SummaryService);
>        summaryService['summaryDataSource'].next({ executing_tasks: [], finished_tasks: [] });
> -      pools = [];
> -      addPool('a');
> -      addPool('b');
> -      addPool('c');
> -      component.pools = pools;
> -      spyOn(poolService, 'getList').and.callFake(() => of(pools));
> -      fixture.detectChanges();
> +      setUpPools();
>      });
>  
>      it('gets all pools without executing pools', () => {
> @@ -241,4 +246,46 @@ describe('PoolListComponent', () => {
>        expect(component.transformPgStatus(pgStatus)).toEqual(expected);
>      });
>    });
> +
> +  describe('getSelectionTiers', () => {
> +    const setSelectionTiers = (tiers: number[]) => {
> +      component.selection.selected = [
> +        {
> +          tiers
> +        }
> +      ];
> +      component.selection.update();
> +      component.getSelectionTiers();
> +    };
> +
> +    it('should select multiple existing cache tiers', () => {
> +      setSelectionTiers([0, 1, 2]);
> +      expect(component.selectionCacheTiers).toEqual(pools);
> +    });
> +
> +    it('should select cache existing tier', () => {
> +      setSelectionTiers([0]);
> +      expect(component.selectionCacheTiers).toEqual([{ pg_num: 256, pool: 0, pool_name: 'a' }]);
> +    });
> +
> +    it('should not select cache tier if id is invalid', () => {
> +      setSelectionTiers([-1]);
> +      expect(component.selectionCacheTiers).toEqual([]);
> +    });
> +
> +    it('should not select cache tier if empty', () => {
> +      setSelectionTiers([]);
> +      expect(component.selectionCacheTiers).toEqual([]);
> +    });
> +
> +    // This test will fail with the current solution
> +    it('should be able to selected one pool with multiple tiers, than with a single tier, than with no tiers', () => {
> +      setSelectionTiers([0, 1, 2]);
> +      expect(component.selectionCacheTiers).toEqual(pools);
> +      setSelectionTiers([0]);
> +      expect(component.selectionCacheTiers).toEqual([{ pg_num: 256, pool: 0, pool_name: 'a' }]);
> +      setSelectionTiers([]);
> +      expect(component.selectionCacheTiers).toEqual([]);
> +    });
> +  });
>  });
> ```
> 
> My solution to fix that would be the following:
> 
> ```diff
> diff --git a/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts b/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
> index d27d198aec..af55f3a12a 100644
> --- a/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
> +++ b/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
> @@ -182,12 +182,6 @@ export class PoolListComponent implements OnInit {
>  
>    getSelectionTiers() {
>      const cacheTierIds = this.selection.hasSingleSelection ? this.selection.first()['tiers'] : [];
> -    this.selectionCacheTiers.splice(0, cacheTierIds.length);
> -
> -    this.pools.forEach((pool) => {
> -      if (cacheTierIds.indexOf(pool.pool) > -1) {
> -        this.selectionCacheTiers.push(pool);
> -      }
> -    });
> +    this.selectionCacheTiers = this.pools.filter((pool) => cacheTierIds.includes(pool.pool))
>    }
>  }
> ```
Very appreciate your advice. That is a mistake.